frame/staking/src/offchain_election.rs

```diff
@@ -155,19 +163,192 @@ pub fn get_balancing_iters<T: Trait>() -> usize {
+				break;
+			},
+			// we found the right value - early exit the function.
+			Ok(next) => return next
+		}
+		step = step / 2;
+		current_weight = weight_with(voters);
+	}
+
+
+	// Time to finish.
+	// We might have reduced less than expected due to rounding error. Reduce linearly one last
+	// time.
+	let mut initialized = false;
+	let mut prev_voters = voters;
+	loop {
+		let next = next_voters(current_weight, voters, 1).unwrap_or(voters);
+		// no change? break.
+		if next == voters {
+			break
+		}
+
+		// then we are stuck in a loop;
+		if prev_voters == next {
+			break;
+		}
+
+		// skip 1 loop, then update these 3 variables in a sequence:
+		// prev_voters --> voter --> next
+		if !initialized {
+			initialized = true
+		} else {
+			prev_voters = voters
+		}
+
+		voters = next;
+		current_weight = weight_with(voters);
+	}
+
+	voters.min(size.nominators)
+}
+
+/// Greedily reduce the size of the a solution to fit into the block, w.r.t. weight.
+///
+/// The weight of the solution is foremost a function of the number of voters (i.e.
+/// `compact.len()`). Aside from this, the other components of the weight are invariant. The number
+/// of winners shall not be changed (otherwise the solution is invalid) and the `ElectionSize` is
+/// merely a representation of the total number of stakers.
+///
+/// Thus, we reside to stripping away some voters. This means only changing the `compact` struct.
+///
+/// Note that the solution is already computed, and the winners are elected based on the merit of
+/// teh entire stake in the system. Nonetheless, some of the voters will be removed further down the
+/// line.
+///
+/// Indeed, the score must be computed **after** this step. If this step reduces the score too much,
+/// then the solution will be discarded.
+pub fn trim_to_weight<T: Trait, FN>(
+	maximum_allowed_voters: u32,
+	mut compact: CompactAssignments,
+	nominator_index: FN,
+) -> Result<CompactAssignments, OffchainElectionError>
+where
+	for<'r> FN: Fn(&'r T::AccountId) -> Option<NominatorIndex>,
+{
+	if let Some(to_remove) = compact.len().checked_sub(maximum_allowed_voters as usize) {
+		// grab all voters and sort them by least stake.
+		let mut voters_sorted = <Nominators<T>>::iter()
+			.map(|(who, _)| {
+				(
+					who.clone(),
+					<Module<T>>::slashable_balance_of_vote_weight(&who),
+				)
+			})
+			.collect::<Vec<_>>();
+		voters_sorted.sort_by_key(|(_, y)| *y);
+
+		// start removing from the least stake. Iterate until we know enough have been removed.
+		let mut removed = 0;
+		for (maybe_index, _stake) in voters_sorted
+			.iter()
+			.map(|(who, stake)| (nominator_index(&who), stake))
+		{
+			let index = maybe_index.ok_or(OffchainElectionError::NominatorSnapshotCorrupt)?;
+			if compact.remove_voter(index) {
+				crate::log!(
+					debug,
+					"removed a voter at index {} with stake {:?} from compact to reduce the size",
+					index,
+					_stake,
+				);
+				removed += 1
+			}
+
+			if removed > to_remove {
```

Comment:
shouldn't it be `>=` ?
if you have 2 to remove you will remove right now 3.
Also you can do on the iterator `.take(to_remove)`

Reply:
indeed, this was off by one. 

> Also you can do on the iterator .take(to_remove)

This was considered and I didn't do it because as you see inside the loop, an iteration might not remove anyone. So we don't know how far we have to iterate, so `.take()` might cause the resulting iterator to be too few.